File: deeptech_semantic_search/jobs/regenerate_expert_embedding.py
"""
Regenerate Embedding for Single Expert
Utility script to update embedding for a specific expert

Usage: python jobs/regenerate_expert_embedding.py <expertId>
"""

import asyncio
import sys
import os
from datetime import datetime

# Add parent directory to path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from services.embedding_service import EmbeddingService
from services.database_service import DatabaseService

async def regenerate_expert_embedding(expert_id: str):
    """Regenerate embedding for a single expert"""
    print('🔄 Regenerating expert embedding...\n')
    print(f'Expert ID: {expert_id}\n')

    try:
        # Initialize services
        embedding_service = EmbeddingService()
        database_service = DatabaseService()

        # 1. Initialize embedding service
        print('📦 Loading AI model...')
        await embedding_service.initialize()
        print('✅ Model loaded\n')

        # 2. Initialize database connection
        await database_service.initialize()

        # 3. Fetch expert from database
        print('🔍 Fetching expert data...')
        expert = await database_service.get_expert_for_embedding(expert_id)

        if not expert:
            print(f'❌ Expert not found with ID: {expert_id}')
            print('Please check the expert ID and try again.\n')
            return None

        expert_name = expert.get('name', 'Unknown')
        print(f'✅ Found expert: {expert_name}\n')

        # 4. Build text representation
        print('📝 Building text representation...')
        expert_text = embedding_service.build_expert_text(expert)

        if not expert_text.strip():
            print('❌ Expert has no text content to embed')
            print('Profile appears to be empty. Please add bio, skills, or domains.\n')
            return None

        print(f'✅ Text length: {len(expert_text)} characters')
        print(f'Preview: "{expert_text[:100]}..."\n')

        # 5. Generate new embedding
        print('🧠 Generating embedding...')
        embedding = await embedding_service.generate_embedding(expert_text)
        print(f'✅ Generated {len(embedding)}-dimensional embedding\n')

        # 6. Update in database
        print('💾 Updating database...')
        result = await database_service.update_expert_embedding(expert_id, embedding, expert_text)

        print('✅ Successfully updated embedding!')
        print(f'   Expert: {expert_name}')
        print(f'   Updated at: {datetime.now()}')
        print(f'   Text length: {len(expert_text)} characters')
        print(f'   Embedding dimensions: {len(embedding)}\n')

        return result

    except Exception as error:
        print(f'❌ Error: {str(error)}')
        raise error

def main():
    """Main entry point"""
    if len(sys.argv) != 2:
        print('Usage: python jobs/regenerate_expert_embedding.py <expertId>')
        print('Example: python jobs/regenerate_expert_embedding.py abc-123-def')
        sys.exit(1)

    expert_id = sys.argv[1]
    asyncio.run(regenerate_expert_embedding(expert_id))

if __name__ == "__main__":
    main()

File: deeptech_backend/jobs/regenerateExpertEmbedding.js
/**
 * Regenerate Embedding for Single Expert
 * Utility script to update embedding for a specific expert
 * 
 * Usage: node jobs/regenerateExpertEmbedding.js <expertId>
 */

import embeddingService from '../services/embeddingService.js';
import expertModel from '../models/expertModel.js';

async function regenerateExpertEmbedding(expertId) {
  console.log('🔄 Regenerating expert embedding...\n');
  console.log(`Expert ID: ${expertId}\n`);
  
  try {
    // 1. Initialize embedding service
    console.log('📦 Loading AI model...');
    await embeddingService.initialize();
    console.log('✅ Model loaded\n');
    
    // 2. Fetch expert from database
    console.log('🔍 Fetching expert data...');
    const expert = await expertModel.getById(expertId);
    
    if (!expert) {
      console.error(`❌ Expert not found with ID: ${expertId}`);
      console.error('Please check the expert ID and try again.\n');
      process.exit(1);
    }
    
    const expertName = `${expert.first_name || ''} ${expert.last_name || ''}`.trim() || 'Unknown';
    console.log(`✅ Found expert: ${expertName}\n`);
    
    // 3. Build text representation
    console.log('📝 Building text representation...');
    const expertText = embeddingService.buildExpertText({
      bio: expert.experience_summary,
      skills: expert.skills,
      domains: expert.domains,
      expertise_areas: expert.expertise_areas
    });
    
    if (!expertText.trim()) {
      console.error('❌ Expert has no text content to embed');
      console.error('Profile appears to be empty. Please add bio, skills, or domains.\n');
      process.exit(1);
    }
    
    console.log(`✅ Text length: ${expertText.length} characters`);
    console.log(`Preview: "${expertText.substring(0, 100)}..."\n`);
    
    // 4. Generate embedding
    console.log('🤖 Generating embedding...');
    const startTime = Date.now();
    const embedding = await embeddingService.generateEmbedding(expertText);
    const duration = Date.now() - startTime;
    
    console.log(`✅ Embedding generated (${duration}ms)`);
    console.log(`   Dimensions: ${embedding.length}`);
    console.log(`   Sample values: [${embedding.slice(0, 3).map(v => v.toFixed(4)).join(', ')}...]\n`);
    
    // 5. Update database
    console.log('💾 Saving to database...');
    const result = await expertModel.updateEmbedding(expertId, embedding, expertText);
    
    console.log('✅ Database updated successfully\n');
    
    // 6. Summary
    console.log('='.repeat(60));
    console.log('🎉 EMBEDDING REGENERATION COMPLETE!');
    console.log('='.repeat(60));
    console.log(`Expert ID:        ${expertId}`);
    console.log(`Expert Name:      ${expertName}`);
    console.log(`Updated At:       ${result.embedding_updated_at}`);
    console.log(`Embedding Size:   ${embedding.length} dimensions`);
    console.log(`Text Length:      ${expertText.length} characters`);
    console.log(`Processing Time:  ${duration}ms`);
    console.log('='.repeat(60) + '\n');
    
    console.log('✅ Expert is now ready for semantic search!\n');
    
    return result;
    
  } catch (error) {
    console.error('\n💥 ERROR:', error.message);
    console.error(error.stack);
    process.exit(1);
  }
}

// Command line interface
if (import.meta.url === `file://${process.argv[1]}`) {
  const expertId = process.argv[2];
  
  if (!expertId) {
    console.error('❌ Error: Expert ID is required\n');
    console.log('Usage:');
    console.log('  node jobs/regenerateExpertEmbedding.js <expertId>\n');
    console.log('Example:');
    console.log('  node jobs/regenerateExpertEmbedding.js 7f8c3a2e-1b4d-4c9a-8e2f-3d5a6b7c8d9e\n');
    process.exit(1);
  }
  
  regenerateExpertEmbedding(expertId)
    .then(() => process.exit(0))
    .catch(error => {
      console.error('Fatal error:', error);
      process.exit(1);
    });
}

export default regenerateExpertEmbedding;
